## Importing the data

In [4]:
import pandas as pd

df = pd.read_csv('data/credit-card.csv')
df.head()

CUST_ID      BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0  C10001    40.900749           0.818182      95.40              0.00   
1  C10002  3202.467416           0.909091       0.00              0.00   
2  C10003  2495.148862           1.000000     773.17            773.17   
3  C10004  1666.670542           0.636364    1499.00           1499.00   
4  C10005   817.714335           1.000000      16.00             16.00   

   INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                    95.4      0.000000             0.166667   
1                     0.0   6442.945483             0.000000   
2                     0.0      0.000000             1.000000   
3                     0.0    205.788017             0.083333   
4                     0.0      0.000000             0.083333   

   ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                    0.000000                          0.083333   
1                    0.000000                          0.000000   
2                    1.000000                          0.000000   
3                    0.083333                          0.000000   
4                    0.083333                          0.000000   

   CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                0.000000                 0              2        1000.0   
1                0.250000                 4              0        7000.0   
2                0.000000                 0             12        7500.0   
3                0.083333                 1              1        7500.0   
4                0.000000                 0              1        1200.0   

      PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
0   201.802084        139.509787          0.000000      12  
1  4103.032597       1072.340217          0.222222      12  
2   622.066742        627.284787          0.000000      12  
3     0.000000               NaN          0.000000      12  
4   678.334763        244.791237          0.000000      12

## Preprocessing the data

Right of the bet, there are columns we don't need. Let's drop them already

In [5]:
df.drop(columns = ['CUST_ID', 'TENURE'], inplace = True)
df.head()

BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0    40.900749           0.818182      95.40              0.00   
1  3202.467416           0.909091       0.00              0.00   
2  2495.148862           1.000000     773.17            773.17   
3  1666.670542           0.636364    1499.00           1499.00   
4   817.714335           1.000000      16.00             16.00   

   INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                    95.4      0.000000             0.166667   
1                     0.0   6442.945483             0.000000   
2                     0.0      0.000000             1.000000   
3                     0.0    205.788017             0.083333   
4                     0.0      0.000000             0.083333   

   ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                    0.000000                          0.083333   
1                    0.000000                          0.000000   
2                    1.000000                          0.000000   
3                    0.083333                          0.000000   
4                    0.083333                          0.000000   

   CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                0.000000                 0              2        1000.0   
1                0.250000                 4              0        7000.0   
2                0.000000                 0             12        7500.0   
3                0.083333                 1              1        7500.0   
4                0.000000                 0              1        1200.0   

      PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  
0   201.802084        139.509787          0.000000  
1  4103.032597       1072.340217          0.222222  
2   622.066742        627.284787          0.000000  
3     0.000000               NaN          0.000000  
4   678.334763        244.791237          0.000000

It's also important to look for missing data

In [6]:
missing = df.isna().sum()
missing

BALANCE                               0
BALANCE_FREQUENCY                     0
PURCHASES                             0
ONEOFF_PURCHASES                      0
INSTALLMENTS_PURCHASES                0
CASH_ADVANCE                          0
PURCHASES_FREQUENCY                   0
ONEOFF_PURCHASES_FREQUENCY            0
PURCHASES_INSTALLMENTS_FREQUENCY      0
CASH_ADVANCE_FREQUENCY                0
CASH_ADVANCE_TRX                      0
PURCHASES_TRX                         0
CREDIT_LIMIT                          1
PAYMENTS                              0
MINIMUM_PAYMENTS                    313
PRC_FULL_PAYMENT                      0
dtype: int64

For this case, instead of dropping the entire *MINIMUM_PAYMENTS* column, we're going to fill those missing values with the median value

In [7]:
df.fillna(df.median(), inplace=True)
missing = df.isna().sum()
missing

BALANCE                             0
BALANCE_FREQUENCY                   0
PURCHASES                           0
ONEOFF_PURCHASES                    0
INSTALLMENTS_PURCHASES              0
CASH_ADVANCE                        0
PURCHASES_FREQUENCY                 0
ONEOFF_PURCHASES_FREQUENCY          0
PURCHASES_INSTALLMENTS_FREQUENCY    0
CASH_ADVANCE_FREQUENCY              0
CASH_ADVANCE_TRX                    0
PURCHASES_TRX                       0
CREDIT_LIMIT                        0
PAYMENTS                            0
MINIMUM_PAYMENTS                    0
PRC_FULL_PAYMENT                    0
dtype: int64

### Normalizing the data

In [8]:
from sklearn import preprocessing

min_max_scaler = preprocessing.MinMaxScaler()
np_df = min_max_scaler.fit_transform(df) # fit_transform transforms our dataframe and returns returns a numpy array
df = pd.DataFrame(np_df, columns=df.columns) # Converting to a dataframe

df.describe()

BALANCE  BALANCE_FREQUENCY    PURCHASES  ONEOFF_PURCHASES  \
count  8950.000000        8950.000000  8950.000000       8950.000000   
mean      0.082154           0.877271     0.020457          0.014534   
std       0.109306           0.236904     0.043570          0.040722   
min       0.000000           0.000000     0.000000          0.000000   
25%       0.006736           0.888889     0.000808          0.000000   
50%       0.045864           1.000000     0.007367          0.000932   
75%       0.107868           1.000000     0.022637          0.014166   
max       1.000000           1.000000     1.000000          1.000000   

       INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
count             8950.000000   8950.000000          8950.000000   
mean                 0.018270      0.020766             0.490351   
std                  0.040193      0.044491             0.401371   
min                  0.000000      0.000000             0.000000   
25%                  0.000000      0.000000             0.083333   
50%                  0.003956      0.000000             0.500000   
75%                  0.020828      0.023629             0.916667   
max                  1.000000      1.000000             1.000000   

       ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
count                 8950.000000                       8950.000000   
mean                     0.202458                          0.364437   
std                      0.298336                          0.397448   
min                      0.000000                          0.000000   
25%                      0.000000                          0.000000   
50%                      0.083333                          0.166667   
75%                      0.300000                          0.750000   
max                      1.000000                          1.000000   

       CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
count             8950.000000       8950.000000    8950.000000   8950.000000   
mean                 0.090096          0.026413       0.041089      0.148390   
std                  0.133414          0.055485       0.069435      0.121491   
min                  0.000000          0.000000       0.000000      0.000000   
25%                  0.000000          0.000000       0.002793      0.051753   
50%                  0.000000          0.000000       0.019553      0.098497   
75%                  0.148148          0.032520       0.047486      0.215359   
max                  1.000000          1.000000       1.000000      1.000000   

          PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  
count  8950.000000       8950.000000       8950.000000  
mean      0.034170          0.011058          0.153715  
std       0.057078          0.030531          0.292499  
min       0.000000          0.000000          0.000000  
25%       0.007556          0.002236          0.000000  
50%       0.016894          0.004088          0.000000  
75%       0.037482          0.010322          0.142857  
max       1.000000          1.000000          1.000000

## Clustering

In [9]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=5, n_init=10, max_iter=300)
pred_labels = kmeans.fit_predict(df)
pred_labels

array([0, 0, 2, ..., 1, 0, 2])

Before we can proceed to interpret any meaning from the clusters, we should make sure that the clustering made is actually good.

## Validating the clusters

### The Silhouette Score

This score can vary between -1 and 1, and the higher the value the better the clusters are defined

**Advantages**
- The defined range between -1 and 1 makes it easier to understand how well defined the clusters are (a value of 0 indicates that the clusters are overlapping)
- Dense and separated clusters gives a higher value, which relates to a standard concept of a cluster

**Drawbacks**
- The Silhouette Coefficient is generally higher for convex clusters than other concepts of clusters, such as density based clusters like those obtained through DBSCAN.
- High computational complexity: O(n^2)

In [10]:
from sklearn.metrics import silhouette_score

s_score = silhouette_score(df, pred_labels)
s_score

0.3538287529295635

Our Silhouette Score was 0.3538, which is somewhat good, but not great. Let's see how changing the number of clusters will affect the silhouette score!

In [11]:
n_clusters_values = [i for i in range(2, 11)] # [2, 3, 4, 5, 6, 7, 8, 9, 10]
s_scores = []

for n_clusters in n_clusters_values:
    cluster = KMeans(n_clusters=n_clusters)
    labels = cluster.fit_predict(df)

    s_scores.append(silhouette_score(df, labels))

s_scores

[0.4134923417305637,
 0.4023268704570294,
 0.35387231279146025,
 0.3538513564213388,
 0.35925536278656167,
 0.33896052310398045,
 0.3609246746316273,
 0.3677592930611809,
 0.3388186789406923]

In [15]:
import plotly.graph_objects as go

fig = go.Figure()
fig.add_trace(go.Scatter(
    x = n_clusters_values,
    y = s_scores
))
fig.update_layout(
    xaxis_title = 'Number of clusters',
    yaxis_title = 'Silhouette Coefficient'
)
fig.show()

### The Davies-Bouldin Score

This index signifies the average ‘similarity’ between clusters, where the similarity is a measure that compares the distance between clusters with the size of the clusters themselves. A lower Davies-Bouldin index relates to a model with better separation between the clusters

**Advantages**
- The computation of Davies-Bouldin is simpler than that of Silhouette scores.
- The index is computed only quantities and features inherent to the dataset.

**Drawbacks**
- The usage of centroid distance limits the distance metric to Euclidean space.
- The Davies-Boulding index is generally higher for convex clusters than other concepts of clusters, such as density based clusters like those obtained from DBSCAN.

In [13]:
from sklearn.metrics import davies_bouldin_score

dbs = davies_bouldin_score(df, pred_labels)
dbs

1.1814548335015707

In [14]:
n_clusters_values = [i for i in range(2, 11)] # [2, 3, 4, 5, 6, 7, 8, 9, 10]
dbs_scores = []

for n_clusters in n_clusters_values:
    cluster = KMeans(n_clusters=n_clusters)
    labels = cluster.fit_predict(df)

    dbs_scores.append(davies_bouldin_score(df, labels))

dbs_scores

[1.0391965514303874,
 1.2544263159796005,
 1.3087646269910276,
 1.1818712154511037,
 1.1728828477055468,
 1.1566262990930454,
 1.0551369146421414,
 0.9963781694507708,
 1.055519951483303]

In [16]:
import plotly.graph_objects as go

fig = go.Figure()
fig.add_trace(go.Scatter(
    x = n_clusters_values,
    y = dbs_scores
))
fig.update_layout(
    xaxis_title = 'Number of clusters',
    yaxis_title = 'Davies-Bouldin Score'
)
fig.show()

### Calinski-Harabasz Score

The Calinski-Harabasz index also known as the Variance Ratio Criterion, is the ratio of the sum of between-clusters dispersion and of inter-cluster dispersion for all clusters, the higher the score , the better the performances.

**Advantages**
- The score is higher when clusters are dense and well separated, which relates to a standard concept of a cluster.
- Fast to compute.

**Drawbacks**
- The Calinski-Harabasz index is generally higher for convex clusters than other concepts of clusters, such as density based clusters like those obtained through DBSCAN.

In [17]:
from sklearn.metrics import calinski_harabasz_score

calinski = calinski_harabasz_score(df, pred_labels)
calinski

4505.628186074391

In [18]:
n_clusters_values = [i for i in range(2, 11)] # [2, 3, 4, 5, 6, 7, 8, 9, 10]
calinski_scores = []

for n_clusters in n_clusters_values:
    cluster = KMeans(n_clusters=n_clusters)
    labels = cluster.fit_predict(df)

    calinski_scores.append(calinski_harabasz_score(df, labels))

calinski_scores

[7083.983008245804,
 5409.681892222902,
 4778.804695337414,
 4505.631556410567,
 4273.915206041556,
 4125.327688737381,
 4019.6812694813625,
 4003.47947435482,
 3843.046916265117]

In [19]:
import plotly.graph_objects as go

fig = go.Figure()
fig.add_trace(go.Scatter(
    x = n_clusters_values,
    y = calinski_scores
))
fig.update_layout(
    xaxis_title = 'Number of clusters',
    yaxis_title = 'Caliski-Harabasz Score'
)
fig.show()

Great, we now have the scores for these three important methods o evaluating the quality of a clustering! However, how can we tell these numbers are actually good? What can we interpret from them?

For the Silhouette Coefficient, the closer the number is to **1**, the better. For the Davies-Bouldin, the closer to **0** the better. And for the Calinski-Harabasz, we want it to be **as high as possible**. It's hard to do any sort of evaluating using only these absolute values.

Knowing that, if we take a look at our graphs we can see clearly how 2 clusters is the optimal number of clusters for all the three metrics. So, from now on, **we'll be using 2 clusters instead of 3**. 

In [20]:
def clustering_algorithm(n_clusters, dataset):
    # Creating the clustering
    kmeans = KMeans(n_clusters=n_clusters, n_init=10, max_iter=300)
    labels = kmeans.fit_predict(dataset)

    # Metrics to evaluate the quality of the clustering created
    s = silhouette_score(df, labels)
    dbs = davies_bouldin_score(df, labels)
    calinski = calinski_harabasz_score(df, labels)

    return s, dbs, calinski

In [35]:
s, dbs, calinski = clustering_algorithm(2, df)
print('For 2 clusters: \nSilhouette = {}\nDavies-Bouldin = {}\nCalinski-Harabasz = {}\n'.format(s, dbs, calinski))

For 2 clusters: 
Silhouette = 0.41347358541358686
Davies-Bouldin = 1.0391671364279806
Calinski-Harabasz = 7083.979615179482



Another way of doing the validation is comparing the values we got from our dataset to those of a random dataset, to make sure that ours is better suited to our dataset.

Let's start by creating a new dataset with the same structure as our own

In [36]:
import numpy as np

random_data = np.random.rand(df.shape[0], df.shape[1]) # Creates a dataframe of shape (8950, 16) with random values from 0 to 1
s_random, dbs_random, calinski_random = clustering_algorithm(3, random_data)
print('Random dataset: \nSilhouette = {}\nDavies-Bouldin = {}\nCalinski-Harabasz = {}\n'.format(s_random, dbs_random, calinski_random))
print('Our dataset: \nSilhouette = {}\nDavies-Bouldin = {}\nCalinski-Harabasz = {}\n'.format(s, dbs, calinski))

Random dataset: 
Silhouette = -0.0034483883623445143
Davies-Bouldin = 93.49824816750663
Calinski-Harabasz = 1.115996921113782

Our dataset: 
Silhouette = 0.41347358541358686
Davies-Bouldin = 1.0391671364279806
Calinski-Harabasz = 7083.979615179482



We can see a huge difference between the scores, which likely means our metrics are well fitted!